Remove `validateDeclForNameLookup` and related logic

xymus opened the pull request:
`validateDeclForNameLookup` was a phase executed before `validateDecl` on some decls to avoid cycles. We already moved its work to be lazier, so we don't need this phase anymore. Calling `validateDecl` is enough as it relies on the new lazy services to avoid cycles.

The next goal is to make lazy more of the logic in `validateDecl` so it is evaluated only when it is actually needed.

xymus committed:
Move generic signature debugging code that triggered extra work too early

Fixes failures on circular references.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented:
I believe you can remove `{LazyResolver,TypeChecker}::resolveProtocolEnvironment()` now, too.

xymus committed:
Remove resolveProtocolEnvironment, it should be no more needed

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

xymus committed:
TypeChecker: substMemberTypeWithBase respects structural type validation

xymus committed:
TypeChecker: resolveTypeDecl use the structural type of alias decls

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4647,11 +4541,6 @@ static void finalizeType(TypeChecker &TC, NominalTypeDecl *nominal) {
-
-  if (auto PD = dyn_cast<ProtocolDecl>(nominal)) {
-    for (auto *inherited : PD->getInheritedProtocols())
-      TC.requestNominalLayout(inherited);
```
We still need this part for now -- it triggers synthesis of accessors

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -969,9 +971,18 @@ static Type resolveTypeDecl(TypeDecl *typeDecl, SourceLoc loc,
+  auto aliasDecl = dyn_cast<TypeAliasDecl>(typeDecl);
+  if (resolution.getStage() == TypeResolutionStage::Structural &&
+      aliasDecl && !aliasDecl->getGenericParams() &&
+      !aliasDecl->getUnderlyingTypeLoc().wasValidated()) {
```
There's a similar check in getStructuralType() in GenericSignatureBuilder.cpp. Maybe TypeAliasDecl::getStructuralType() should automatically return the checked underlying type if its available.

xymus replied:
I've applied the suggested change to TypeAliasDecl::getStructuralType. I'll leave this check here because we still want to execute TypeChecker::resolveTypeInContext if the type was previously validated, it provides better diagnostics.

slavapestov commented on lib/Sema/TypeChecker.h:
```diff
@@ -892,8 +890,11 @@ class TypeChecker final : public LazyResolver {
   static Type substMemberTypeWithBase(ModuleDecl *module, TypeDecl *member,
-                                      Type baseTy, bool useArchetypes = true);
+                                      Type baseTy, bool useArchetypes = true,
+                                      TypeResolutionStage stage = TypeResolutionStage::Interface);
```
Maybe if you pass in a TypeResolution instead of a TypeResolutionStage, you can combine the useArchetypes and state parameters into one; useArchetypes then just becomes resolution.usesArchetypes().

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3402,20 +3415,18 @@ Type TypeChecker::substMemberTypeWithBase(ModuleDecl *module,
-    }
+  Type memberType;
+  if (aliasDecl) {
+    memberType = aliasDecl->getUnderlyingTypeLoc().getType();
```
This duplicates the above check, see my comment about the GSB and getStructuralType()

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -988,9 +999,10 @@ static Type resolveTypeDecl(TypeDecl *typeDecl, SourceLoc loc,
-                                      generic);
+  if (!type)
+    type =
+      TypeChecker::resolveTypeInContext(typeDecl, foundDC, resolution,
```
Should getting the structural type be pushed down into resolveTypeInContext(), since it has access to the `resolution` object?

xymus replied:
We probably could, but we also need to avoid the call to the previous `resolveDeclSignature` because it triggers `validateDecl`.

slavapestov reviewed (approved):
A few suggestions you might want to look at. The only important one is the removal of the requestNominalLayout() call.

xymus committed:
TypeAliasDecl::getStructuralType returns a validated type if available

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus committed:
Sema: StructuralTypeRequest wraps the resolved type in a TypeAliasType

This fixes an issue with protocol compositions using a typealias of
anoter protocol composition causing a difference between the flat
structural type and the nested underlying TypeRepr.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1035,5 +1035,21 @@ swift::StructuralTypeRequest::evaluate(Evaluator &evaluator,
+  auto type = resolution.resolveType(typeRepr, options);
+
+  // Wrap ProtocolCompositionTypes in a TypeAliasType
+  if (type->getKind() == TypeKind::ProtocolComposition) {
```
type->is<ProtocolCompositionType>()

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1035,5 +1035,21 @@ swift::StructuralTypeRequest::evaluate(Evaluator &evaluator,
+    Type parent;
+    auto parentDC = D->getDeclContext();
+    if (parentDC->isTypeContext())
+      parent = parentDC->getDeclaredInterfaceType();
```
getSelfInterfaceType()

slavapestov commented on lib/Sema/TypeChecker.h:
```diff
@@ -1024,7 +1021,12 @@ class TypeChecker final : public LazyResolver {
   virtual void resolveDeclSignature(ValueDecl *VD) override {
-    validateDeclForNameLookup(VD);
+    auto dc = VD->getDeclContext();
+    if (auto nominal = dyn_cast<NominalTypeDecl>(dc)) {
```
This should not be necessary?

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

It passed the source compat suite, now trying again with one less hack.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus committed:
Sema: validateDecl validates the context of GenericTypeParamDecls

xymus committed:
Sema: print type alias underlying types when debugging generics

With the recent change to StructuralTypeRequest where all returned types
are wrapped in a TypeAliasType, printing the requirement signature
could display requirements such as `Self.P3Assoc == Self.P3Assoc`.
This change will make sure that when in debug mode, the underlying type
will be displayed instead on the RHS.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus committed:
Remove validateDeclForNameLookup and related logic

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test
